### DISCUSSION TITLE
How to expose constexpr inside std::optional to SWIG?

### DISCUSSION THREAD

**Discussion by SmrtSmrtnik** [swig/swig#2707] [Q&A]
TLDR: trying to write a SWIG .i file with typemap that gives me the ability to translate between  `std::optional<T>` and `java.util.Optional<T>` for all primitives, strings, and a variety of custom classes.

====== original opening post below this line, latest state in this comment: https://github.com/swig/swig/discussions/2707#discussioncomment-7624200 =======
Hello, beginner here. Chances are I am doing something silly.

### What I am trying to do: 
write a typemap for going between std::optional<T> and java.util.Optional<T> for all primitives, strings, and a variety of custom classes.

### What I am leveraging, being new to SWIG:
 the typemap for a custom Optional class linked by @vadz here: https://github.com/swig/swig/issues/1307

### What I have so far, SWIG interface file (excuse the many comments, it's helping me learn SWIG/JNI):
```
%module java_optional

%{
#include <optional>
%}
%include <optional>

%define DEFINE_OPTIONAL_SIMPLE(OptT, T)

namespace std {

  // Use reference, not pointer, typemaps for member variables of this type.
  %naturalvar optional< T >;

  // Allows us to use templates which pass in to DEFINE_OPTIONAL_SIMPLE as OptT (something to name it) and T
  %template(OptT) optional< T >;


  // Note the use of "jboxtype" instead of just "jstype": for primitive types,
  // such as "double", they're different and we must use the former as
  // Optional<> can only be used with reference types in Java.
  %typemap(jstype) optional< T >, const optiona< T >& "java.util.Optional<$typemap(jboxtype, T)>" 

  // This typemap is used for function arguments of this type.
  %typemap(javain,
  //The 'pre' attribute contains code that is generated before the JNI call 
  // and the 'post' attribute contains code generated after the JNI call. 
  //
  // $javainput used in "javain" typemaps.  analogous to the $input special variable. It is replaced by the parameter name.
          pre="    OptT opt$javainput = $javainput.isPresent() ? new OptT($javainput.get()) : new OptT();",
  // The 'pgcppname' attribute is used to change the premature garbage collection prevention parameter
  // name passed to the JNI function. This is sometimes needed when the 'pre' typemap creates a temporary
  // variable which is then passed to the JNI function
          pgcppname="opt$javainput")
  // $javaclassname similar to $1_type (first argument Real C/C++ datatype). Expands to class name for use
  // in Java given a pointer. If the type does not have an associated proxy class, it expands to the type
  // wrapper class name.
          const optional< T >& "$javaclassname.getCPtr(opt$javainput)"

  // This typemap is for functions returning objects of this type.
  %typemap(javaout) optional< T >, const optionals< T >& {
      OptT optvalue = new OptT($jnicall, $owner);
      if (optvalue.IsValid())
        return java.util.Optional.of(optvalue.Get());
      else
        return java.util.Optional.empty();
  }

}
%enddef

DEFINE_OPTIONAL_SIMPLE("Opt_double" ,double)  

%inline %{
#include <iostream>
  std::optional<double> optional_double_test(double x) { 
    if (x)
    {
      std::cout << "C++ " << *x << std::endl;
      return std::optional<double>(x);  //Optional.of(x)
    }
    else
    {
      std::cout << "C++ " << "NULL" << std::endl;
      return std::optional<double>(); //Optional.isEmpty()
    }
  }
%}
```

### What I have so far, Java file to exercise it with:
```
public class Main
{
    static
    {
        System.loadLibrary("java_optional");
    }

    public static void main(String[] args) throws Exception 
    {               
        System.out.println("java (empty should be true): " + java_optional.optional_double_test(null).isEmpty());
        System.out.println("java double: " + java_optional.optional_double_test(123.456).get());        
    }
}
```

When I run ` swig  -v -DSWIGWORDSIZE64 -c++ -java java_optional.i `
I get error message `java_optional.i:10: Error: Unable to find 'optional'`  (in my local file, line number 10 is the `%include <optional>` line)

My local installation has the standard libraries all in /usr/include/c++/10 so I try `  swig  -v -DSWIGWORDSIZE64 -c++ -java -I/usr/include/c++/10 java_optional.i ` but get the same error message.

If I leave out the ``%include <optional>`, it just complains the first time I try to invoke the macro, at ` DEFINE_OPTIONAL_SIMPLE("Opt_double" ,double)  `  that ` java_optional.i:58: Error: Template 'optional' undefined. java_optional.i:58: Error: Syntax error in input(1). `.  It is my understanding that I need to include it anyway though, for use by SWIG itself rather than preprocessor. Regardless, for some reason, the command line swig seems to not see the library.  I am using latest SWIG master built locally, since I wanted the large integer fix that's not in any released version yet, hence the `-DSWIGWORDSIZE64` option, not that that makes a difference here.

So what am I missing?  Been googling for a couple of days now and can't see anything like this, so it's probably a me problem where I forgot something simple. Many thanks.

### EDIT: So I suspect I know what's happening, but I'm not sure how to solve it.
I made a .cpp file with contents of the example from https://en.cppreference.com/w/cpp/utility/optional, that is to say
```
#include <iostream>
#include <optional>
#include <string>
 
// optional can be used as the return type of a factory that may fail
std::optional<std::string> create(bool b)
{
    if (b)
        return "Godzilla";
    return {};
}
 
// std::nullopt can be used to create any (empty) std::optional
auto create2(bool b)
{
    return b ? std::optional<std::string>{"Godzilla"} : std::nullopt;
}
 
int main()
{
    std::cout << "create(false) returned "
              << create(false).value_or("empty") << '\n';
 
    // optional-returning factory functions are usable as conditions of while and if
    if (auto str = create2(true))
        std::cout << "create2(true) returned " << *str << '\n';
}
```

And I ran `g++ optionalsinc.cpp -fPIC` on it, and got tons of errors, but one of them was a warning that std::optional is a C++ 17 feature.
So I tried `g++ -std=c++17 optionalsinc.cpp -fPIC` and that compiled fine.  If I add `-o optionalsinc.exe`, it executes fine too, with the right outputs.
I did ` g++ -dM -E -x c++  /dev/null | grep -F __cplusplus` and that gives out `#define __cplusplus 201402L` so apparently my g++ is defaulting to c++14 instead of 17.

So... I think I need to tell command line SWIG to use 17 (guessing it is also defaulting to 14, and then not being as clear about the error message as g++ was).  Er... how do I do that? I do not see it as a command line option on `swig -help`.

... Nevermind, because it's not just <optional> that refuses to import. Any std library I try to pull in, I get the same kind of error message about it.

**Comment by ojwb** [comment]
> how do I instruct command line SWIG to use the -std=c++17 compilation flag?

This question doesn't make sense - SWIG generates C++ code, but it doesn't compile the code for you so it doesn't use any compilation flags.

I think your problem just boils down to SWIG not having support for `std::optional` currently - see #1307.

**Reply by SmrtSmrtnik** [reply]
Hi @ojwb ,
Ok it is possible I'm just misunderstanding how SWIG operates.  I understand that there is no built in support for `std::optional` in SWIG currently, however that very link #1307 has links to having a custom Optional class typemap into Java's Optional and the suggestion that converting it to work with std::optional should be relatively easy.  The code I posted up top as java_optional.i, I believe should do it.  But instead command line SWIG tells me it doesn't know what `optional` in `%include <optional>` refers to.

It makes the same complaint if for example I try to do `%include <iostream>` or any other std library.  I need to make SWIG typemaps (as I understand them) for a variety of std objects produced by the library (i.e. std::chrono::duration into java.time.Duration) I've been tasked to convert to Java via SWIG, and `std::optional` just happens to be the first one up, but if I can't extend its functionality then the rest of them will be just as much of a non-starter.

**Reply by ojwb** [reply]
For SWIG's iostream support you'd use `%include <std_iostream.i>`, but SWIG doesn't currently supply a `std_optional.i`.

Howver if you're supplying your own typemaps for `std::optional` then you don't need to `%include` anything here.  It's like if you were using a C++ < 17 compiler and wanted a `std::optional` like class you could supply your own but you wouldn't try to `#include <optional>` as well.

**Reply by SmrtSmrtnik** [reply]
Yes, that's what I'm trying to do, supply my own typemaps for `std::optional` (and eventually some others, but that's step 2). 
If I don't `%include <optional>`, I just get a complaint from command line SWIG on the `DEFINE_OPTIONAL_SIMPLE("Opt_double" ,double) ` line instead, saying it doesn't know what optional is.  That's why I thought I had to `%include <optional>`.  Do you know why that might be? The reason I macroed it is so I can make templates out for all the primitives + structs and classes I expect to come over inside std::optional<T>s.  In the sample code I have above I just have double, since if it works for that one it *should* work for others as well.

**Reply by ojwb** [reply]
You don't seem to define the `optional` template for SWIG - equivalent is in https://gist.github.com/vadz/1d7a7c1be4f55fe45b46ea06454349e1 in the code you seem to be working from:

```
template <typename T>
class OptionalValue
```
[...]

**Reply by SmrtSmrtnik** [reply]
I could be wrong here, but I thought that was a custom in-house `OptionalValue` class, not the `std::optional` from standard library.

So from what I understand of SWIG, if I have some code I wrote in foo.cpp and foo.h, I can `%include "foo.h"` in a foo.i file and command line SWIG will then wrap it so I can make calls to foo's methods from either elsewhere in the .i file or from eventual Java code, correct?  And that's an alternative to just retyping up the declarations.

That's why I thought I had to `%include <optional>`, for that same reason.  In the https://gist.github.com/vadz/1d7a7c1be4f55fe45b46ea06454349e1 they define a `OptionalValue` class but just the declarations, not implementation.  Then, `%include "java/optional.i"` afterwards for Java (or others for the other languages), and java/optional.i is referencing this `OptionalValue` class.  

Ok so to give an example on another track, let's say I am writing a .i file to be the typemap for going back and forth between `java.time.Duration` and `std::chrono::duration`.  In the .i file, I would need to either retype the declarations for `std::chrono::duration` OR `%import <chrono>`, correct?

Maybe I should just retype the declarations for `std::optional`?

**Reply by SmrtSmrtnik** [reply]
Made some progress, but still not quite there. The following compiles and links fine for `std::optional<double>` only. Once I have that fully working, I will be turning it into a macro for other types.  However, trying to then compile the Java code and it has issues finding the relevant symbols, so not quite there yet.

## java_optional.i:
```
%module java_optional

//these are just in case, probably/possibly needed
%include <stdint.i>
%include <typemaps.i>
%include <cpointer.i>

%{
#include </usr/local/include/c++/9.5.0/optional>
%}
%include </usr/local/include/c++/9.5.0/optional>

namespace std {
  template<typename T> class optional {};
}

namespace std {

  // Use reference, not pointer, typemaps for member variables of this type.
  %naturalvar optional< double >;

  %template(optional_double) optional< double >;

  // Note the use of "jboxtype" instead of just "jstype": for primitive types,
  // such as "double", they're different and we must use the former as
  // Optional<> can only be used with reference types in Java.
  %typemap(jstype) optional< double >, const optional< double >& "java.util.Optional<$typemap(jboxtype, double)>" 

  // This typemap is used for function arguments of this type.
  %typemap(javain,
          pre="    optional_double opt$javainput = $javainput.isPresent() ? new optional_double($javainput.get()) : new optional_double();",
          pgcppname="opt$javainput")

    const optional< double >& "$javaclassname.getCPtr(opt$javainput)"

    %typemap(javaout) optional< double >, const optional< double >& {
      optional_double optvalue = new optional_double($jnicall, $owner);
      if (optvalue.IsValid())
        return java.util.Optional.of(optvalue.Get());
      else
        return java.util.Optional.empty();
  }

}

%{
#include "usedoubleoptional.h"
%}
%include "usedoubleoptional.h"

```

## usedoubleoptional.h
```
#include </usr/local/include/c++/9.5.0/iostream>
#include </usr/local/include/c++/9.5.0/optional>

std::optional<double> optional_double_test(double* x) {
    std::optional<double> retval; //empty 
    if (x)
    {
      std::cout << "C++ " << *x << std::endl;
      retval = std::optional<double>(*x);
    }
    else
    {
      std::cout << "C++ " << "NULL" << std::endl;      
    }
    return retval;
  }

std::optional<double> optional_double_test(double x) {
  return optional_double_test(&x);
}

void printOptionalFromJava(std::optional<double> opt) {  
   std::cout << "C++ " << (opt == std::nullopt ? "opt dbl is NULL" : "opt dbl is good") << std::endl;
}
```

`swig  -v -DSWIGWORDSIZE64 -c++ -java java_optional.i` runs with no complaints.
`g++ -c -std=c++17 usedoubleoptional.h java_optional_wrap.cxx -I/opt/jdk-17/include -I/opt/jdk-17/include/linux -fPIC` runs with no complaints
`g++ -shared java_optional_wrap.o -o libjava_optional.so` runs with no complaints


So now with
## Main.java
```
import java.util.Optional;

public class Main
{
    static
    {
        System.loadLibrary("java_optional");
    }

    public static void main(String[] args) throws Exception 
    {       
       Optional<Double> shouldBeEmpty = java_optional.optional_double_test(null);
        System.out.println("java (empty should be true): " + shouldBeEmpty.isEmpty());
        Optional<Double> shouldHaveValue = java_optional.optional_double_test(123.456D);
        System.out.println("java double: " + shouldHaveValue.get());   
        
        Optional<Double> emptyIn = Optional.empty();
        java_optional.printOptionalFromJava(emptyIn);
        Optional<Double> filledIn = Optional.of(1.234d);
        java_optional.printOptionalFromJava(filledIn);   
    }
}
```

`javac -Xdiags:verbose Main.java` says
```
./java_optional.java:14: error: cannot find symbol
      if (optvalue.IsValid())
                  ^
  symbol:   method IsValid()
  location: variable optvalue of type optional_double
./java_optional.java:15: error: cannot find symbol
        return java.util.Optional.of(optvalue.Get());
                                             ^
  symbol:   method Get()
  location: variable optvalue of type optional_double
./java_optional.java:23: error: cannot find symbol
      if (optvalue.IsValid())
                  ^
  symbol:   method IsValid()
  location: variable optvalue of type optional_double
./java_optional.java:24: error: cannot find symbol
        return java.util.Optional.of(optvalue.Get());
                                             ^
  symbol:   method Get()
  location: variable optvalue of type optional_double
./java_optional.java:30: error: method getCPtr in class optional_double cannot be applied to given types;
    java_optionalJNI.printOptionalFromJava(optional_double.getCPtr(opt), opt);
                                                          ^
  required: optional_double
  found:    Optional<Double>
  reason: argument mismatch; Optional<Double> cannot be converted to optional_double
5 errors
```

If I look at SWIG generated `java_optional.java` it is
```
/* ----------------------------------------------------------------------------
 * This file was automatically generated by SWIG (https://www.swig.org).
 * Version 4.2.0
 *
 * Do not make changes to this file unless you know what you are doing - modify
 * the SWIG interface file instead.
 * ----------------------------------------------------------------------------- */

public class java_optional implements java_optionalConstants {
  public static java.util.Optional<Double> optional_double_test(SWIGTYPE_p_double x) {
      // // OptT optvalue = new OptT(java_optionalJNI.optional_double_test__SWIG_0(SWIGTYPE_p_double.getCPtr(x)), false);
      optional_double optvalue = new optional_double(java_optionalJNI.optional_double_test__SWIG_0(SWIGTYPE_p_double.getCPtr(x))
, false);
      if (optvalue.IsValid())
        return java.util.Optional.of(optvalue.Get());
      else
        return java.util.Optional.empty();
  }

  public static java.util.Optional<Double> optional_double_test(double x) {
      // // OptT optvalue = new OptT(java_optionalJNI.optional_double_test__SWIG_1(x), false);
      optional_double optvalue = new optional_double(java_optionalJNI.optional_double_test__SWIG_1(x), false);
      if (optvalue.IsValid())
        return java.util.Optional.of(optvalue.Get());
      else
        return java.util.Optional.empty();
  }

  public static void printOptionalFromJava(java.util.Optional<Double> opt) {
    java_optionalJNI.printOptionalFromJava(optional_double.getCPtr(opt), opt);
  }

}
```

So seems like it's not sure what optional_double.IsValid() or optional_double.Get() is supposed to do?
I feel like I'm so close.  As I look `optional_double.java`, it's clear that it does not have implementation or even declaration of IsValid(), or Get(), so that's probably step 1.They are not in https://en.cppreference.com/w/cpp/utility/optional, they are in the custom class I had originally built on top of! I need to use has_value() and value(), respectively.  After making that change, javac complains that those cannot be found either.  And looking at SWIG generated `optional_double.java` it still does not have those in there either.  So I assume I need to declare it in `java_optional.i` somewhere?  Doesn't 
```
%{
#include </usr/local/include/c++/9.5.0/optional>
%}
%include </usr/local/include/c++/9.5.0/optional>
```
take care of that for me though?!
Is it because both `has_value()` and `value()` are declared in `<optional>` as `constexpr`? Specifically
```
constexpr bool has_value() const noexcept
{ return this->_M_is_engaged(); }

constexpr const _Tp& value() const&
{
    return this->_M_is_engaged()
          ? this->_M_get()
          : (__throw_bad_optional_access(), this->_M_get());
}
```
Per SWIG documentation, `SWIG parses and identifies the keyword constexpr, but cannot fully utilise it. These C++ compile time constants are usable as runtime constants from the target languages. `. Is there some way I can expose them?
